### ISSUE TITLE
Placeholder

### ISSUE THREAD

**Issue by OpenModelica-TracImporter** [OpenModelica/OpenModelica#6342]
This is a placeholder created during migration to preserve original issue numbers.

----

Migrated-From: https://trac.openmodelica.org/OpenModelica/ticket/6342

**Comment by OpenModelica-TracImporter** [comment]
*Comment by **Francesco Casella** on Sat, 16 Jan 2021 17:03:10 GMT:*

See [PR 7083](https://github.com/OpenModelica/OpenModelica/pull/7083).

**Comment by OpenModelica-TracImporter** [comment]
*Comment by **Francesco Casella** on Sat, 16 Jan 2021 19:09:04 GMT:*

OK, things are a bit more complicated than I thought.

On one hand, there is probably an incentive at using the builtin/dense Newton/homotopy solver for initialization, as it has a number of heuristics and tricks to handle convergence failures that Kinsol does not have. On the other hand, in case there are large algebraic loops in the simulation equations, the best solver is IDA in daeMode (see also #6340), so this option won't be used during simulation, as daeMode can only work with IDA/Kinsol and needs to use a sparse solver for systems with more than a few dozens variables (i.e. practically all meaningful Modelica models).

So, it makes sense to increase the default minimium size for the nonlinear solver to a larger size; even though the dense solver is less efficient, it is only used once at initialiation, and there is a premium at having increased robustness. On the other hand, initialization should not take an unreasonable amount of time. Given the results shown in the table, I think a reasonable compromise is to set `nlssMinSize=100` by default.

There is no such issue in the case of linear solvers; in fact, if we turn tearing off by default (see #6196 and [PR 6969](https://github.com/OpenModelica/OpenModelica/pull/6969)) it will be essential not to use the dense solver inappropriately. So, I'll leave the limit to 10 there. In fact, we could even set it to a lower value.

I updated [PR 7083](https://github.com/OpenModelica/OpenModelica/pull/7083) accordingly.

**Comment by OpenModelica-TracImporter** [comment]
*Comment by **Francesco Casella** on Sat, 16 Jan 2021 19:13:19 GMT:*

There is another issue, though, which is demonstrated in the updated test package, `DefaultSolverReducedMinSize` package. Even if I set `-nlssMinSize=10`, the builtin Newton/homotopy solver is still used; apparently, since it only works with a dense linear solver, it ignores that flag.

In order to get the sparse nonlinear solver to do its job, I need to set `-nls="kinsol"`, and at that point, apparently, it uses the sparse linear solver by default anyway.

This is all a bit messy, I guess we should clean this up for 1.17.0 if possible. We can discuss this during the next devmeeting.

**Comment by OpenModelica-TracImporter** [comment]
*Comment by **Francesco Casella** on Wed, 14 Jul 2021 13:15:19 GMT:*

I just updated the attached sample library, which now includes tests for both linear and nonlinear systems.

I confirm that with the default settings, sparse solvers are never selected, even in cases where it would be absolutely obvious that it is advantageous to do so.

**Comment by OpenModelica-TracImporter** [comment]
*Comment by **Francesco Casella** on Wed, 14 Jul 2021 14:05:31 GMT:*

I made some further enquiry with the attached test suite. For example, if I run `Linear.DefaultSolver.Test_N_80`, it uses Lapack, i.e. the default.

If I then add `-lssMinSize=1` I get this interesting output

```
Maximum system size for using linear sparse solver changed to 1

initialize linear system solvers
1 linear systems

Using sparse solver for linear system 0,
because density of 0.037 remains under threshold of 0.200 and size of 80 exceeds threshold of 1.

The maximum density and the minimal system size for using sparse solvers can be specified
using the runtime flags '<-lssMaxDensity=value>' and '<-lssMinSize=value>'.

Start solving Linear System 161 (size 80) at time 0 with Klu Solver
```

So it seems that the logic to select the sparse solver is:

if size > lssMinSize **and** density < lssMaxDensity then use sparse else use dense

I don't find this very useful. I mean, `lssMinSize` is set by default to 4001. I don't think it is reasonable to use a dense solver when you have 4000 unknowns: the jacobian would take 16 M elements, that is to say 128 MB of RAM. If the density is below the threshold, we should use a sparse solver **regardless of the size**. If the size is above the threshold, we should use a sparse solver anyway, to try minimizing the memory footprint.

So, to my understanding, the selection logic should be:

if size > lssMinSize **or** density < lssMaxDensity then use sparse else use dense

Ditto for nonlinear solvers

if size > nlssMinSize **or** density < nlssMaxDensity then use sparse else use dense

We could then trim the defaults a bit based on some experiments, but we should change the logic right away. The current one often caused some of my models to become impossibly slow, and besides, without giving any clue so as to the cause. They were simply trying to factorize a dense matrix with some million zeros in it.

**Comment by OpenModelica-TracImporter** [comment]
*Comment by **Francesco Casella** on Fri, 16 Jul 2021 00:20:32 GMT:*

Done in [PR 7695](https://github.com/OpenModelica/OpenModelica/pull/7695).

`@phannebohm`, can you please port this to v.18 maintenance as well.

**Comment by OpenModelica-TracImporter** [comment]
*Comment by **Francesco Casella** on Fri, 26 Nov 2021 09:07:59 GMT:*

Issue eventually resolved in 1.19.0

**Comment by OpenModelica-TracImporter** [comment]
*Comment by **Francesco Casella** on Fri, 26 Nov 2021 09:08:10 GMT:*

<sub>Trac change: resolution → fixed, status new → closed</sub>
